### ISSUE TITLE
Rule proposal: `no-instanceof-function`

### ISSUE THREAD

**Issue by dimaMachina** [sindresorhus/eslint-plugin-unicorn#2157]
### Description

similar to https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-instanceof-array.md

### Fail

```js
foo instanceof Function
```

### Pass

```js
typeof foo === 'function'
```

**Comment by B2o5T** [comment]
another case but for `instanceof Object`

## Fail

```js
let foo = {}
foo instanceof Object // true

let bar = null
bar instanceof Object // false
```

## Pass

```js
Boolean(foo) && typeof foo === 'object'
Boolean(bar) && typeof bar === 'object'
```

**Comment by silverwind** [comment]
Would make it a generic `no-instanceof` with default config like `["Array", "Function"]`.

**Comment by sindresorhus** [comment]
> Would make it a generic no-instanceof with default config like ["Array", "Function"].

👍

**Comment by sindresorhus** [comment]
Accepted.

**Comment by zanminkian** [comment]
@dimaMachina From my point of view, `foo installceof Object` is better than `Boolean(foo) && typeof foo === 'object'`. People usually forget to add `Boolean(foo)`. Moreover, non-object judgement will be verbose, like `if (!(Boolean(foo) && typeof foo === 'object')){}` vs `if (!(foo instanceof Object))`. Obviously, the `instanceof` wins.

**Comment by sindresorhus** [comment]
`instanceof` should be avoided for built-ins as it does not work across realms (iframes, Node.js VM, etc).

**Comment by zanminkian** [comment]
> `instanceof` should be avoided for built-ins as it does not work across realms (iframes, Node.js VM, etc).

Thanks! `instanceof` seems not safe in some edge case. Personally，I seldom use `instanceof`. I think we should add a new rule `no-instanceof`. The default option is `all`. User can config its options to `['Array','Function','Object']` manually.
